### PR TITLE
JBPM-7867 - kie-server-spring-boot springboot 2 support

### PR DESCRIPTION
**Pull request by mswiderski**
@MarianMacik would you mind to run Kie Server tests via spring boot? Or maybe it's time to have Jenkins job for that...

### CODE REVIEW

**Comment by mswiderski** [comment]
depends on https://github.com/kiegroup/droolsjbpm-knowledge/pull/363

File: kie-spring-boot/kie-spring-boot-autoconfiguration/jbpm-spring-boot-data-sources/src/main/java/org/jbpm/springboot/datasources/JBPMDataSourceAutoConfiguration.java [hunk 1]
```diff
@@ -138,11 +142,16 @@ private XADataSource createXaDataSourceInstance(String className) {
...
+        values.put("user", properties.determineUsername());
+        values.put("password", properties.determinePassword());
+        values.put("url", properties.determineUrl());
+        values.putAll(properties.getXa().getProperties());
+        
+        ConfigurationPropertyNameAliases aliases = new ConfigurationPropertyNameAliases();
+        aliases.addAliases("user", "username");
+        
+        Binder binder = new Binder(values.withAliases(aliases));
+        binder.bind(ConfigurationPropertyName.EMPTY, Bindable.ofInstance(target));
```
Comment (MarianMacik) [inline kie-spring-boot/kie-spring-boot-autoconfiguration/jbpm-spring-boot-data-sources/src/main/java/org/jbpm/springboot/datasources/JBPMDataSourceAutoConfiguration.java]: This EMPTY means it will apply all properties possible?

File: kie-spring-boot/kie-spring-boot-autoconfiguration/kie-server-spring-boot-autoconfiguration/pom.xml [hunk 2]
```diff
@@ -89,6 +89,16 @@
...
       </exclusions>
     </dependency>
+    <dependency>
+      <groupId>org.springframework.security</groupId>
+      <artifactId>spring-security-core</artifactId>
+      <exclusions>
+        <exclusion>
+          <groupId>aopalliance</groupId>
+          <artifactId>aopalliance</artifactId>
+        </exclusion>
```
Comment (MarianMacik) [inline kie-spring-boot/kie-spring-boot-autoconfiguration/kie-server-spring-boot-autoconfiguration/pom.xml]: Any reason for exclusion?

File: kie-spring-boot/kie-spring-boot-samples/jbpm-spring-boot-sample-basic/src/main/java/org/jbpm/springboot/samples/WebSecurityConfig.java [hunk 3]
```diff
@@ -33,8 +36,15 @@ protected void configure(HttpSecurity http) throws Exception {
...
+        PasswordEncoder encoder = PasswordEncoderFactories.createDelegatingPasswordEncoder();
+              
         auth.inMemoryAuthentication()
-                .withUser("john").password("john1").roles("jbpm,HR,IT,Accounting,PM");
+                .withUser("john").password(encoder.encode("john1")).roles("jbpm,HR,IT,Accounting,PM");
+    }
+    
+    @Bean
+    public PasswordEncoder passwordEncoder() {
+        return PasswordEncoderFactories.createDelegatingPasswordEncoder();
```
Comment (MarianMacik) [inline kie-spring-boot/kie-spring-boot-samples/jbpm-spring-boot-sample-basic/src/main/java/org/jbpm/springboot/samples/WebSecurityConfig.java]: As I understand it now, this bean declaration is not needed since we don't inject it anywhere. I tried it without it and I was able to log in correctly.

File: kie-spring-boot/kie-spring-boot-autoconfiguration/jbpm-spring-boot-autoconfiguration/pom.xml [hunk 4]
```diff
@@ -73,6 +73,12 @@
     <dependency>
       <groupId>org.jbpm</groupId>
       <artifactId>jbpm-kie-services</artifactId>
+      <exclusions>
+        <exclusion>
+          <groupId>javax.validation</groupId>
+          <artifactId>validation-api</artifactId>
+        </exclusion>
+      </exclusions>
```
[the remark this answers is not in the document]
Comment (mswiderski) [reply kie-spring-boot/kie-spring-boot-autoconfiguration/jbpm-spring-boot-autoconfiguration/pom.xml]: this is brought by kie-soup-dataset-api and it has 1.0 version while spring boot web requires 2.1 so excluding it here as it will be added by spring anyway

File: kie-spring/src/test/resources/org/kie/spring/persistence/persistence_var_beans_env.xml [hunk 5]
```diff
@@ -61,9 +61,9 @@
     </kie:object-marshalling-strategies>
   </kie:environment>
 
-  <kie:environment id="env2">
-    <kie:entity-manager-factory ref="myEmf"/>
+  <kie:environment id="env2">    
     <kie:transaction-manager ref="txManager"/>
+    <kie:entity-manager-factory ref="myEmf"/>
```
[the remark this answers is not in the document]
Comment (mswiderski) [reply kie-spring/src/test/resources/org/kie/spring/persistence/persistence_var_beans_env.xml]: it's logical as first txManager is created and then entity manager that relies on it

File: kie-spring-boot/kie-spring-boot-samples/kie-server-spring-boot-integ-tests-sample/src/main/java/org/kie/server/springboot/samples/IntegrationTestsWebSecurityConfig.java [hunk 6]
```diff
@@ -56,10 +58,19 @@ public HttpFirewall customStrictHttpFirewall() {
...
+        .withUser("Administrator").password(encoder.encode("usetheforce123@")).roles("kie-server", "guest", "Administrators")
+        .and()        
+        .withUser("john").password(encoder.encode("usetheforce123@")).roles("kie-server", "guest", "engineering", "HR", "IT", "Accounting")
+        .and()        
+        .withUser("mary").password(encoder.encode("usetheforce123@")).roles("kie-server", "guest", "engineering", "HR", "IT", "Accounting");
+    }
+    
+    @Bean
+    public PasswordEncoder passwordEncoder() {
+        return PasswordEncoderFactories.createDelegatingPasswordEncoder();
```
[the remark this answers is not in the document]
Comment (mswiderski) [reply kie-spring-boot/kie-spring-boot-samples/kie-server-spring-boot-integ-tests-sample/src/main/java/org/kie/server/springboot/samples/IntegrationTestsWebSecurityConfig.java]: fixed

File: kie-spring-boot/kie-spring-boot-autoconfiguration/jbpm-spring-boot-data-sources/src/main/java/org/jbpm/springboot/datasources/JBPMDataSourceAutoConfiguration.java (hunk 1 above)
Comment (mswiderski) [reply kie-spring-boot/kie-spring-boot-autoconfiguration/jbpm-spring-boot-data-sources/src/main/java/org/jbpm/springboot/datasources/JBPMDataSourceAutoConfiguration.java]: that's how I understand it as well.

File: kie-spring-boot/kie-spring-boot-autoconfiguration/kie-server-spring-boot-autoconfiguration/pom.xml (hunk 2 above)
Comment (mswiderski) [reply kie-spring-boot/kie-spring-boot-autoconfiguration/kie-server-spring-boot-autoconfiguration/pom.xml]: duplicated classes

File: kie-spring-boot/kie-spring-boot-samples/jbpm-spring-boot-sample-basic/src/main/java/org/jbpm/springboot/samples/WebSecurityConfig.java (hunk 3 above)
Comment (mswiderski) [reply kie-spring-boot/kie-spring-boot-samples/jbpm-spring-boot-sample-basic/src/main/java/org/jbpm/springboot/samples/WebSecurityConfig.java]: fixed

**Comment by mswiderski** [comment]
Jenkins retest this

**Comment by MarianMacik** [comment]
kie-spring:
```
Duplicate classes found:
  Found in:
    org.springframework:spring-jcl:jar:5.1.3.RELEASE:compile
    org.slf4j:jcl-over-slf4j:jar:1.7.25:compile
  Duplicate classes:
    org/apache/commons/logging/Log.class
    org/apache/commons/logging/impl/NoOpLog.class
    org/apache/commons/logging/impl/SimpleLog.class
    org/apache/commons/logging/LogFactory.class
```

kie-server-spring-boot-autoconfiguration-drools module has the same.

I think we can get rid of org.slf4j:jcl-over-slf4j:jar dependency in kie-spring/pom.xml, since spring-core has its own slf4j bridge.

File: kie-spring-boot/kie-spring-boot-samples/jbpm-spring-boot-sample-basic/pom.xml [hunk 7]
```diff
@@ -30,7 +30,7 @@
   <dependencies>
     <dependency>
       <groupId>org.kie</groupId>
-      <artifactId>jbpm-spring-boot-starter-basic</artifactId>
+      <artifactId>jbpm-spring-boot-starter-basic</artifactId>
```
Comment (MarianMacik) [inline kie-spring-boot/kie-spring-boot-samples/jbpm-spring-boot-sample-basic/pom.xml]: When fixing deps, you can also delete this unnecessary whitespace.

**Comment by mswiderski** [comment]
Jenkins retest this

File: kie-spring-boot/pom.xml [hunk 8]
```diff
@@ -33,18 +33,21 @@
   </modules>
 
   <properties>
-    <version.spring-boot>1.5.12.RELEASE</version.spring-boot>
-    <version.cxf.jaxrs>3.1.11</version.cxf.jaxrs>
-    <version.me.snowdrop.narayana>1.0.1</version.me.snowdrop.narayana>
+    <version.org.springframework>5.1.3.RELEASE</version.org.springframework>
+    <version.org.spring-boot>2.1.1.RELEASE</version.org.spring-boot>
```
Comment (mareknovotny) [inline kie-spring-boot/pom.xml]: there is 2.1.2.RELEASE which seems is in our certified version https://github.com/snowdrop/spring-boot-bom/blob/sb-2.1.x/pom.xml#L176. Was that an omission or intent to not use that version?
